File: sdk/go/flusher.go
// flusher.go — background goroutine that periodically drains the buffer and
// sends events to the gateway. Using a single goroutine (not a pool) keeps
// delivery ordered and simplifies the transport interface.
package watcher

import (
	"log"
	"sync"
	"time"
)

// flusher owns the goroutine that drains the buffer on a timer tick or
// when the buffer reaches the flushAt threshold.
type flusher struct {
	interval  time.Duration
	flushAt   int
	buf       *buffer
	transport Transport

	trigger chan struct{} // signals the goroutine to flush immediately
	done    chan struct{} // signals the goroutine to stop after a final flush
	wg      sync.WaitGroup
}

// newFlusher creates a flusher and starts its background goroutine immediately.
func newFlusher(buf *buffer, transport Transport, interval time.Duration, flushAt int) *flusher {
	f := &flusher{
		interval:  interval,
		flushAt:   flushAt,
		buf:       buf,
		transport: transport,
		trigger:   make(chan struct{}, 1), // buffered so push never blocks
		done:      make(chan struct{}),
	}
	f.wg.Add(1)
	go f.run()
	return f
}

// run is the flush loop. It drains on the interval tick, on an early trigger
// (flushAt threshold), and once more when stop() closes the done channel.
func (f *flusher) run() {
	defer f.wg.Done()
	ticker := time.NewTicker(f.interval)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			f.flush()
		case <-f.trigger:
			f.flush()
		case <-f.done:
			// Final flush — ensures no events are silently dropped on shutdown.
			f.flush()
			return
		}
	}
}

// notify is called after every push. If the buffer has reached the flushAt
// threshold it signals the goroutine to flush now rather than wait for the
// next timer tick. The channel is buffered (capacity 1) so this never blocks.
func (f *flusher) notify() {
	if f.buf.size() >= f.flushAt {
		select {
		case f.trigger <- struct{}{}:
		default: // goroutine already has a pending trigger queued
		}
	}
}

// flush drains the buffer and sends all events in a single batch.
// Errors are logged but never returned — telemetry must not crash the app.
func (f *flusher) flush() {
	events := f.buf.drain()
	if len(events) == 0 {
		return
	}
	if err := f.transport.Send(events); err != nil {
		log.Printf("watcher: flush error (%d events dropped): %v", len(events), err)
	}
}

// stop signals the goroutine to perform a final flush and exit, then waits
// for it to finish. Returns an error if already stopped.
func (f *flusher) stop() error {
	select {
	case <-f.done:
		// Already closed — calling stop twice is a no-op error.
		return nil
	default:
		close(f.done)
		f.wg.Wait()
		return nil
	}
}

File: sdk/rust/src/flusher.rs
// flusher.rs — background thread that drains the buffer on a timer or trigger.
//
// Uses a single std::sync::mpsc channel carrying FlushSignal. The thread
// blocks with recv_timeout(flush_interval) — the timeout IS the timer tick.
// A Flush signal causes an early drain; Stop causes a final flush then exits.
//
// A single thread (not a pool) keeps delivery ordered and the transport simple.
use std::sync::{mpsc, Arc};
use std::thread;
use std::time::Duration;

use crate::buffer::Buffer;
use crate::transport::Transport;

/// Signals sent from the main thread to the flusher thread.
pub(crate) enum FlushSignal {
    /// Drain and send now (triggered when flushAt threshold is reached).
    Flush,
    /// Drain, send, then exit.
    Stop,
}

/// Owns the flusher thread handle and the sender half of the signal channel.
pub(crate) struct Flusher {
    handle: Option<thread::JoinHandle<()>>,
    tx: mpsc::SyncSender<FlushSignal>,
}

impl Flusher {
    /// Spawns the flusher thread. The thread holds the only reference to the
    /// transport so no locking is needed around transport calls.
    pub fn start(
        buffer: Arc<Buffer>,
        transport: Arc<dyn Transport>,
        flush_interval: Duration,
        flush_at: usize,
    ) -> Self {
        // Bounded channel of capacity 1 — a second trigger while one is already
        // queued is a no-op (try_send fails silently). Prevents trigger storms.
        let (tx, rx) = mpsc::sync_channel::<FlushSignal>(1);

        let handle = thread::Builder::new()
            .name("watcher-flusher".to_owned())
            .spawn(move || run(rx, buffer, transport, flush_interval, flush_at))
            .expect("watcher: failed to spawn flusher thread");

        Flusher { handle: Some(handle), tx }
    }

    /// Signals the background thread to flush immediately (non-blocking).
    /// If a signal is already queued, this is a no-op.
    pub fn notify(&self, buffer_len: usize, flush_at: usize) {
        if buffer_len >= flush_at {
            let _ = self.tx.try_send(FlushSignal::Flush);
        }
    }

    /// Sends Stop, waits for the final flush, and joins the thread.
    pub fn stop(mut self) {
        let _ = self.tx.send(FlushSignal::Stop);
        if let Some(h) = self.handle.take() {
            let _ = h.join();
        }
    }
}

fn run(
    rx: mpsc::Receiver<FlushSignal>,
    buffer: Arc<Buffer>,
    transport: Arc<dyn Transport>,
    flush_interval: Duration,
    _flush_at: usize,
) {
    loop {
        match rx.recv_timeout(flush_interval) {
            Ok(FlushSignal::Stop) | Err(mpsc::RecvTimeoutError::Disconnected) => {
                // Final flush before exiting so no buffered events are lost.
                flush_once(&buffer, &*transport);
                return;
            }
            Ok(FlushSignal::Flush) | Err(mpsc::RecvTimeoutError::Timeout) => {
                flush_once(&buffer, &*transport);
            }
        }
    }
}

/// Drains the buffer and sends all events in one batch.
/// Logs errors to stderr but never panics — telemetry must not crash the app.
fn flush_once(buffer: &Buffer, transport: &dyn Transport) {
    let events = buffer.drain();
    if events.is_empty() {
        return;
    }
    if let Err(e) = transport.send(&events) {
        eprintln!("watcher: flush error ({} events dropped): {e}", events.len());
    }
}
